readthedocs/oauth/services/bitbucket.py

```diff
@@ -0,0 +1,210 @@
+        """Sync repositories and teams from Bitbucket API"""
+        if sync:
+            self.sync_repositories()
+            self.sync_teams()
+
+    def sync_repositories(self):
+        """Sync repositories from Bitbucket API"""
+        # Get user repos
+        try:
+            repos = self.paginate(
+                'https://bitbucket.org/api/2.0/repositories/?role=member')
+            for repo in repos:
+                self.create_repository(repo)
+        except (TypeError, ValueError) as e:
+            log.error('Error syncing Bitbucket repositories: %s',
+                      str(e), exc_info=True)
+            raise Exception('Could not sync your Bitbucket repositories, '
+                            'try reconnecting your account')
+
+        # Because privileges aren't returned with repository data, run query
+        # again for repositories that user has admin role for, and update
+        # existing repositories.
+        try:
+            resp = self.paginate(
+                'https://bitbucket.org/api/2.0/repositories/?role=admin')
+            repos = (
+                RemoteRepository.objects
+                .filter(users=self.user,
+                        full_name__in=[r['full_name'] for r in resp],
+                        account=self.account)
+            )
+            for repo in repos:
+                repo.admin = True
+                repo.save()
+        except (TypeError, ValueError):
+            pass
+
+    def sync_teams(self):
+        """Sync Bitbucket teams and team repositories"""
+        try:
+            teams = self.paginate(
+                'https://api.bitbucket.org/2.0/teams/?role=member'
+            )
+            for team in teams:
+                org = self.create_organization(team)
+                repos = self.paginate(team['links']['repositories']['href'])
+                for repo in repos:
+                    self.create_repository(repo, organization=org)
+        except ValueError as e:
+            log.error('Error syncing Bitbucket organizations: %s',
+                      str(e), exc_info=True)
+            raise Exception('Could not sync your Bitbucket team repositories, '
+                            'try reconnecting your account')
+
+    def create_repository(self, fields, privacy=DEFAULT_PRIVACY_LEVEL,
+                          organization=None):
+        """Update or create a repository from Bitbucket API response
+
+        .. note::
+            The :py:data:`admin` property is not set during creation, as
+            permissions are not part of the returned repository data from
+            Bitbucket.
+
+        :param fields: dictionary of response data from API
+        :param privacy: privacy level to support
+        :param organization: remote organization to associate with
+        :type organization: RemoteOrganization
+        :rtype: RemoteRepository
+        """
+        if (fields['is_private'] is True and privacy == 'private' or
```

Comment:
I'm wondering if we should be creating objects for all the returned data, and then filtering based on privacy in the output? That would simplify a lot of the privacy checks in this code, and allow us to maintain valid state for all projects & teams a user has. It just seems like having that privacy logic on import isn't really buying us much, and it's hard to reason about.